x-pack/plugins/apm/common/utils/join_by_key/index.ts

```diff
@@ -23,24 +23,43 @@ import { isEqual } from 'lodash';
 */
 
 type JoinedReturnType<
+  T extends Record<string, any>,
+  U extends UnionToIntersection<T>
+> = Array<
+  Partial<U> &
+    {
+      [k in keyof T]: T[k];
+    }
+>;
+
+export function joinByKey<
   T extends Record<string, any>,
   U extends UnionToIntersection<T>,
   V extends keyof T & keyof U
-> = Array<Partial<U> & Record<V, U[V]>>;
+>(items: T[], key: V): JoinedReturnType<T, U>;
 
 export function joinByKey<
   T extends Record<string, any>,
   U extends UnionToIntersection<T>,
-  V extends keyof T & keyof U
->(items: T[], key: V): JoinedReturnType<T, U, V> {
-  return items.reduce<JoinedReturnType<T, U, V>>((prev, current) => {
+  V extends keyof T & keyof U,
+  W extends JoinedReturnType<T, U>,
+  X extends (a: T, b: T) => ValuesType<W>
+>(items: T[], key: V, mergeFn: X): W;
+
+export function joinByKey(
```

Comment:
Maybe we could add some description to make it a bit more clear what this function does.

Reply:
There is one, top of the file.